File: 01.docker/05.docker-compose/microservices/node/cache-worker/src/worker.js
import dotenv from "dotenv";
import amqp from "amqplib";
dotenv.config();
// ─── In-Memory Cache (replaces Redis) ───────────────────────────────────────
// Simple Map: short_code -> original_url
// Also tracks hit count per code.
const cache = new Map(); // { short_code: { url, hits, cachedAt } }

function cacheSet(short_code, original_url) {
  cache.set(short_code, {
    url: original_url,
    hits: 0,
    cachedAt: new Date().toISOString(),
  });
  console.log(`[Cache] SET ${short_code} -> ${original_url}`);
}

function cacheHit(short_code) {
  if (cache.has(short_code)) {
    cache.get(short_code).hits++;
    console.log(
      `[Cache] HIT  ${short_code} | total hits: ${cache.get(short_code).hits}`
    );
  }
}

// ─── RabbitMQ Consumer ───────────────────────────────────────────────────────
const EXCHANGE = "url.events";
const QUEUE = "cache_q";

async function start(retries = 20) {
  for (let repeater = 0; repeater <= retries; repeater++) {
    try {
      const conn = await amqp.connect(
        process.env.RABBIT_URL || "amqp://guest:guest@rabbitmq:5672"
      );
      const ch = await conn.createChannel();

      await ch.assertExchange(EXCHANGE, "fanout", {
        durable: true,
      });
      await ch.assertQueue(QUEUE, { durable: true });
      await ch.bindQueue(QUEUE, EXCHANGE, "");

      console.log(`[CacheWorker] Listening on queue: ${QUEUE}`);

      ch.consume(QUEUE, (msg) => {
        console.log("message is:", msg);
        // console.log(msg);
        if (!msg) return;
        try {
          const event = JSON.parse(msg.content.toString());
          console.log(event);

          if (event.type === "URL_CREATED") {
            cacheSet(event.short_code, event.original_url);
          } else if (event.type === "URL_VISITED") {
            cacheHit(event.short_code);
          }

          ch.ack(msg);
        } catch (err) {
          console.error(
            "[CacheWorker] Failed to process message:",
            err.message
          );
          ch.nack(msg, false, false);
        }
      });

      // Periodically print cache stats
      setInterval(() => {
        console.log(`[Cache] Stats — ${cache.size} entries:`);
        for (const [code, data] of cache.entries()) {
          console.log(`  ${code}: hits=${data.hits} url=${data.url}`);
        }
      }, 30_000);

      return;
    } catch (error) {
      console.error(
        `[CacheWorker] Connect attempt ${repeater + 1} failed:`,
        error.message
      );
      await new Promise((r) => setTimeout(r, 3000));
    }
  }
  process.exit(1);
}

start().catch(console.error);

File: 01.docker/05.docker-compose/microservices/python/cache-worker/worker.py
"""
Cache Worker (Python)
Subscribes to cache_q on the url.events fanout exchange.
Uses a plain Python dict as in-memory cache (replaces Redis).
"""

import os, json, time
import pika
from dotenv import load_dotenv

load_dotenv()

RABBIT_URL = os.getenv("RABBIT_URL", "amqp://guest:guest@rabbitmq:5672")
EXCHANGE   = "url.events"
QUEUE      = "cache_q"

# ─── In-Memory Cache ─────────────────────────────────────────────────────────
cache: dict[str, dict] = {}   # { short_code: { url, hits, cached_at } }

def cache_set(short_code: str, original_url: str):
    cache[short_code] = {"url": original_url, "hits": 0, "cached_at": time.time()}
    print(f"[Cache] SET  {short_code} -> {original_url}")

def cache_hit(short_code: str):
    if short_code in cache:
        cache[short_code]["hits"] += 1
        print(f"[Cache] HIT  {short_code} | total hits: {cache[short_code]['hits']}")

# ─── Message Handler ─────────────────────────────────────────────────────────
def on_message(ch, method, _props, body):
    try:
        event = json.loads(body)
        if event.get("type") == "URL_CREATED":
            cache_set(event["short_code"], event["original_url"])
        elif event.get("type") == "URL_VISITED":
            cache_hit(event["short_code"])
        ch.basic_ack(delivery_tag=method.delivery_tag)
    except Exception as e:
        print(f"[CacheWorker] Error: {e}")
        ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)

# ─── Boot ────────────────────────────────────────────────────────────────────
def start(retries=10):
    for i in range(retries):
        try:
            params = pika.URLParameters(RABBIT_URL)
            connection = pika.BlockingConnection(params)
            ch = connection.channel()

            ch.exchange_declare(exchange=EXCHANGE, exchange_type="fanout", durable=True)
            ch.queue_declare(queue=QUEUE, durable=True)
            ch.queue_bind(queue=QUEUE, exchange=EXCHANGE)
            ch.basic_consume(queue=QUEUE, on_message_callback=on_message)

            print(f"[CacheWorker] Listening on queue: {QUEUE}")
            ch.start_consuming()
            return
        except Exception as e:
            print(f"[CacheWorker] Attempt {i+1} failed: {e}")
            time.sleep(3)

if __name__ == "__main__":
    start()